src/common/tags.c

```diff
@@ -1038,41 +1038,91 @@ uint32_t dt_tag_get_suggestions(GList **result)
+                                "INSERT INTO memory.taglist (id, count)"
+                                " SELECT S.tagid, COUNT(*)"
+                                "  FROM main.tagged_images AS S"
+                                "  WHERE S.tagid NOT IN memory.darktable_tags"
+                                "  GROUP BY S.tagid",
+                                -1, &stmt, NULL);
+    sqlite3_step(stmt);
+    sqlite3_finalize(stmt);
+  }
+
+  // already attached tags
   DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
-                              "INSERT INTO memory.taglist (id, count)"
-                              " SELECT S.tagid, COUNT(*)"
-                              "  FROM main.tagged_images AS S"
-                              "  WHERE S.tagid NOT IN memory.darktable_tags"
-                              "  GROUP BY S.tagid",
+                              "INSERT INTO memory.similar_tags (tagid)"
+                              " SELECT DISTINCT tagid FROM main.tagged_images"
+                              " WHERE imgid IN main.selected_images"
+                              " AND tagid NOT IN memory.darktable_tags",
                               -1, &stmt, NULL);
   sqlite3_step(stmt);
   sqlite3_finalize(stmt);
 
   const uint32_t nb_selected = dt_selected_images_count();
-  /* Now put all the bits together */
-  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
-                              "SELECT T.name, T.id, MT.count, CT.imgnb, T.flags, T.synonyms"
-                              "  FROM memory.taglist MT"
-                              "  JOIN data.tags T ON MT.id = T.id"
-                              "  LEFT JOIN (SELECT tagid, COUNT(DISTINCT imgid) AS imgnb"
-                              "             FROM main.tagged_images"
-                              "             WHERE imgid IN (SELECT imgid FROM main.selected_images)"
-                              "             GROUP BY tagid) AS CT"
-                              "    ON CT.tagid = MT.id"
-                              "  WHERE T.id NOT IN (SELECT DISTINCT tagid"
-                              "                     FROM (SELECT TI.tagid, COUNT(DISTINCT SI.imgid) AS imgnb"
-                              "                           FROM main.selected_images AS SI"
-                              "                           JOIN main.tagged_images AS TI ON TI.imgid = SI.imgid"
-                              "                           GROUP BY TI.tagid)"
-                              "                           WHERE imgnb = ?1)"
-                              "  AND (T.flags IS NULL OR (T.flags & 1) = 0)"
-                              "  ORDER BY MT.count DESC"
-                              "  LIMIT 500",
+  const char *slist = dt_conf_get_string_const("plugins/lighttable/tagging/recent_tags");
+  char *query = NULL;
+  if(confidence != 100)
+    query = g_strdup_printf("SELECT td.name, tagid2,"
```

Comment:
A nice SQL query! Not sure I'll be able to fully understand it :)

Reply:
It's confusing for me as well. That may help to create a third memory table for this (I've used the two available ones). That would help to visualize the different set of images the query uses. And may be faster as well.

I've try to let some comments but I have found I only added confusion... :(

Reply:
added some comments :)